webui/ec2/EC2.py

```diff
@@ -23,78 +23,120 @@ def _user_add_keys(user):
+            users.append(u)
         except ClientError as ex:
-            print (ex);
+            print(ex)
     else:
         for response in iam.get_paginator('list_users').paginate():
             for user in response.get('Users'):
-                u = _ec2user_to_user(user);
-                _user_add_keys(u);
-                users.append(u);
+                u = _ec2user_to_user(user)
+                _user_add_keys(u)
+                users.append(u)
+
+    return users
+
+
+def get_user_by_key(key_id):
+    users = get_users()
 
-    return users;
+    for user in users:
+        for key in user.keys:
+            if key_id == key.key_id:
+                return user
+    return None
+
+
+def get_key(key_id):
+    users = get_users()
+    for user in users:
+        for key in user.keys:
+            if key_id == key.key_id:
+                return key
+    return None
+
+
+def delete_key(key_id):
+    user = get_user_by_key(key_id)
+    if (user is None):
+        return False
+    if (len(user.keys) == 1):
+        delete_user(user.name)
+        return True
+    else:
+        iam_res = boto3.resource(
+            'iam', aws_access_key_id=provider_conf.EC2['key'],
+            aws_secret_access_key=provider_conf.EC2['secret'])
+        key = iam_res.AccessKey(user.name, key_id)
+        key.delete()
+        return True
 
 
 def delete_user(username):
     iam = boto3.client('iam', aws_access_key_id=provider_conf.EC2['key'],
                        aws_secret_access_key=provider_conf.EC2['secret'])
     iam_res = boto3.resource('iam', aws_access_key_id=provider_conf.EC2['key'],
-                       aws_secret_access_key=provider_conf.EC2['secret'])
+                             aws_secret_access_key=provider_conf.EC2['secret'])
 
-    user = iam_res.User(username)
-    user.load()
+    try:
+        user = iam_res.User(username)
+        user.load()
+    except ClientError as ex:
+        print(ex)
+        return False
 
     for key in user.access_keys.all():
         key.delete()
 
-    for response in iam.get_paginator('list_attached_user_policies').paginate(UserName = username):
+    for response in iam.get_paginator('list_attached_user_policies').paginate(
+            UserName=username):
         for policy in response.get('AttachedPolicies'):
-            user.detach_policy(PolicyArn = policy.get('PolicyArn'))
+            user.detach_policy(PolicyArn=policy.get('PolicyArn'))
     user.delete()
+    return True
+
 
 def create_user(username):
     iam_res = boto3.resource('iam', aws_access_key_id=provider_conf.EC2['key'],
-                       aws_secret_access_key=provider_conf.EC2['secret'])
+                             aws_secret_access_key=provider_conf.EC2['secret'])
     user = iam_res.User(username)
     try:
         user.load()
     except ClientError:
         try:
             user.create(Path="/pcw/auto-generated/")
-        except ClientError as err:
+        except ClientError:
```

Comment:
catching exception inside exception looks bad

Reply:
indeed, but in some cases like here seems necessary, as user.load() seems to raise an exception if the user doesn't exist, therefore we create it in the except. Another way without nested try/except could be checking if user is not None and create it if so...

Reply:
as I mention below - are you sure that it is **only** reason why `user.load()` will raise exception?  What if user exists but you got exception here with whatever another reason ?

Reply:
I get the point.
I didn't found a documentation, where they say which exceptions are possible.
So again, I could use get_user() before and then load...

Reply:
let's do the following merge it like it is and I will provide PR how I think it should look like

Reply:
sorry, didn't notice your comment :/ and there was no autofresh here.
I use now the `iam_res.meta.client.exceptions.NoSuchEntityException` exception if the user doesn't exists.
But plz feel free to improve!! thx